design/ceph-external-cluster.md

```diff
@@ -0,0 +1,105 @@
+
+The CSI driver is agnostic of whether Ceph is running locally or externally. The core requirement of the CSI driver is the list of mons and the keyring with which to connect. This metadata is required whether the cluster is local or external. The Rook operator will need to keep this metadata updated throughout the lifetime of the CSI driver.
+
+The flex driver would also be agnostic of the cluster for the same reasons, but we won’t need to worry about the flex driver going forward.
+
+## Rook-Ceph Cluster CRD
+
+In order for Rook to provide the storage to clients in the local cluster, the CephCluster CRD will be created in order for the operator to provide local management of the external cluster. There are several differences needed for the operator to be aware of an external cluster.
+
+1. Before the CephCluster CRD is created, some metadata must be initialized in local configmaps/secrets to allow the local cluster to manage the external cluster.
+   * FSID, mon endpoint(s), admin keyring
+1. The mon, mgr, and osd daemons will not be managed by the local Rook operator. These daemons must be created and managed by the external cluster.
+1. The operator will make a “best effort” to keep the list of mons updated.
+   * If the mons change in the external cluster, the list of mons must be updated in the local cluster.
+   * The operator will need to query the Ceph status periodically (perhaps every minute). If there is a change to the mons, the operator will update the local configmaps/secrets.
+   * If the local operator fails to see changes to the external mons, perhaps because it is down, the mon list could become stale. In that case, the admin will need to update the list similarly to how it was initialized when the local cluster was first created.
+
+The first bullet point above requires an extra manual configuration step by the cluster admin from what he needs in a typical Rook cluster. The other items above will be handled automatically by the Rook operator. The extra step involves exporting metadata from the external cluster and importing it to the local cluster:
+1. The admin creates a yaml file with the needed resources from the external cluster (ideally we would provide a helper script to help automate this task):
+   * Save the mon list and admin keyring
+1. Load the yaml file into the local cluster
+   * `oc create -f <config.yaml>`
+
+The CephCluster CRD will have a new property “external” to indicate whether the cluster is external. If true, the local operator will implement the described behavior.
+
+```yaml
+kind: CephCluster
+spec:
+  external: true
+```
+
+The mgr modules, including the dashboard, would be running in the external cluster. Any configuration that happens through the dashboard would depend on the orchestration modules in that external cluster.
+
+## Block Storage
+
+With the rook-ceph cluster created, the CSI driver integration will cover the Block (RWO) storage and no additional management is needed.
```

Comment:
My only question here is whether the ``CephBlockPool`` works the same way.. when you instantiate one of these it will still create the pool in teh external cluster, right?  And if it already exists, then that 'create' step is a no-op, so in principle multiple kube clusters could share the same ceph cluster and rbd pool?  I think this is basically a question of whether (1) the image names won't conflict and (2) what happens if one cluster's pool CRD says replica 2 and the other says replica 3.

Reply:
Yes, I would expect all that behavior. The two independent clusters could both connect to the same external ceph cluster to consume the same pool. Admins would need to be careful of conflicts, such as the replica count. The operator will update the replica count on a pool each time it runs through the orchestration. This means if the CRDs have different replication its value could change unintentionally. It really is just the pools that would conflict. The image names wouldn't conflict since they are based on the pvc name.

Reply:
A storageclass represents a Pool of a Ceph cluster. The `StorageClass` is injected by the administrator so the replica size will be either the one set by the cluster (if this is a new pool) or will result in a no-op (because it already exists). So I don't think there will be any conflicts, at least in the CSI definition I don't see any mention of the rep size: https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/csi/example/rbd/storageclass.yaml#L9. So I don't see this as a problem. Am I missing something?

However, if there is an option to set the size of the pool in the storageclass declaration then we must honor it.
I believe each storageclass declaration by an admin must be honour not matter what. So if the pool name exists the create a new pool with a different prefix.

This will prevent one Kubernetes cluster removing its storage class and impacting another Kubernetes's storageclass.

So let's make the pool unique for each Kubernetes cluster request.
Makes sense?

Reply:
The storage class only specifies the name of the pool, so I believe storage classes are independent from this design. Here's what I think will help clarify the approach for the external resources...

The `external: true` is only needed on the cluster crd. When the operator is managing the external cluster, it will treat all of the pools, object store, and cephfs as external. There is no need to set the `external` property on all these CRDs because it is implicit for all resources managed by this cluster CRD. The operator would have these behaviors for this cluster:
- Pool CRDs: The first time a pool is created, create it if it doesn't exist yet. If it already exists, never update its properties since it is owned by the external cluster. If the pool CRD is deleted, do not delete the external pool.
- Filesystem CRD: Check if the external filesystem exists. If it doesn't exist, write an error to the log. If it does exist, there is nothing else to do. 
- ObjectStore CRDs: The first time an object store is created, create it if it doesn't exist yet. If it already exists, don't create the pools. Start the RGW daemons in the local cluster. Do not delete the object store when the CRD is deleted since it is owned by the external cluster.
- ObjectStoreUser CRD: Create the user if it doesn't exist. When the CRD is deleted, don't delete the external user. 
- ObjectBucket CRD (in progress): Create the bucket if it doesn't exist. When the CRD is deleted. the bucket will not be deleted.

If CRDs of the same name are created by multiple clusters, only the official external cluster will be the owner of those resources and will be the authority for their settings and lifetime.

Reply:
Tacking onto 2 responses, to provide my comments here.

> The image names wouldn't conflict since they are based on the pvc name.

The PVC names are unique per kubernetes cluster, there is nothing preventing them from conflicting across clusters at the moment.

The image names should be specialized with some form of `csi-id` that is unique to each CSI driver instance, but negotiated outside the scope of the CSI driver. As, there is nothing unique that I can see in each CSI instance that can help specialize this and retain uniqueness and ensure no conflicts.

My thoughts would be to tack this onto the Ceph cluster/pool as user controlled metadata (if such a construct exists within Ceph (thinking along the lines of xattrs on file)), and using the same to add a new unique ID per CSI instance that will access the pool and pass this generated CSI ID to the CSI instance hence.

OR, (possibly very unpopular) ask the admin to inject the CSI instance id and to ensure it is unique.

> The storage class only specifies the name of the pool, so I believe storage classes are independent from this design.

Yes, this is true, and should remain in the future as well.